File: backend/app/routes/candidates.py
from flask import Blueprint, request

from app.services.candidate_service import CandidateService
from app.utils.errors import AppError
from app.utils.responses import error_response, success_response

candidates_bp = Blueprint("candidates", __name__)


@candidates_bp.errorhandler(AppError)
def handle_app_error(err: AppError):
    return error_response(err.message, err.error_code, err.status_code)


@candidates_bp.route("/upload", methods=["POST"])
def upload_resume():
    resume = request.files.get("resume") or request.files.get("file")
    candidate = CandidateService.upload_resume(resume)
    return success_response(
        data=candidate.to_dict(),
        message="Resume uploaded and processed successfully",
        status=201,
    )


@candidates_bp.route("", methods=["GET"])
def list_candidates():
    items = CandidateService.list_candidates()
    return success_response(data={"candidates": items, "count": len(items)})


@candidates_bp.route("/<int:candidate_id>", methods=["GET"])
def get_candidate(candidate_id: int):
    candidate = CandidateService.get_candidate(candidate_id)
    return success_response(data=candidate.to_dict(include_relations=True))


@candidates_bp.route("/<int:candidate_id>", methods=["DELETE"])
def delete_candidate(candidate_id: int):
    result = CandidateService.delete_candidate(candidate_id)
    return success_response(
        data=result,
        message="Candidate deleted successfully",
    )


@candidates_bp.route("/<int:candidate_id>/request-documents", methods=["POST"])
def request_documents(candidate_id: int):
    result = CandidateService.request_documents(candidate_id)
    return success_response(
        data=result,
        message="Document request message generated",
    )


@candidates_bp.route("/<int:candidate_id>/submit-documents", methods=["POST"])
def submit_documents(candidate_id: int):
    pan = request.files.get("pan")
    aadhaar = request.files.get("aadhaar")
    result = CandidateService.submit_documents(candidate_id, pan, aadhaar)
    return success_response(
        data=result,
        message="Documents uploaded successfully",
    )

File: frontend/src/api/candidates.ts
import type {
  ApiResponse,
  Candidate,
  CandidatesListData,
  RequestDocumentsData,
  SubmitDocumentsData,
  UploadCandidateData,
} from '../types/candidate'
import { apiClient } from './client'

export async function uploadResume(
  file: File,
  onProgress?: (percent: number) => void,
) {
  const formData = new FormData()
  formData.append('resume', file)

  const { data } = await apiClient.post<ApiResponse<UploadCandidateData>>(
    '/candidates/upload',
    formData,
    {
      headers: { 'Content-Type': 'multipart/form-data' },
      onUploadProgress: (event) => {
        if (!event.total || !onProgress) return
        onProgress(Math.round((event.loaded * 100) / event.total))
      },
    },
  )
  return data
}

export async function fetchCandidates() {
  const { data } = await apiClient.get<ApiResponse<CandidatesListData>>('/candidates')
  return data
}

export async function fetchCandidate(id: number) {
  const { data } = await apiClient.get<ApiResponse<Candidate>>(`/candidates/${id}`)
  return data
}

export async function requestDocuments(id: number) {
  const { data } = await apiClient.post<ApiResponse<RequestDocumentsData>>(
    `/candidates/${id}/request-documents`,
  )
  return data
}

export async function deleteCandidate(id: number) {
  const { data } = await apiClient.delete<ApiResponse<{ id: number; email: string; deleted: boolean }>>(
    `/candidates/${id}`,
  )
  return data
}

export async function submitDocuments(id: number, pan?: File, aadhaar?: File) {
  const formData = new FormData()
  if (pan) formData.append('pan', pan)
  if (aadhaar) formData.append('aadhaar', aadhaar)

  const { data } = await apiClient.post<ApiResponse<SubmitDocumentsData>>(
    `/candidates/${id}/submit-documents`,
    formData,
    { headers: { 'Content-Type': 'multipart/form-data' } },
  )
  return data
}
